File: src/main/java/no/utgdev/diy/DIY.java
package no.utgdev.diy;

// TODO Oppgave 2
@SuppressWarnings("ALL")
public class DIY {

    public static DIY getInstance() {
        return getInstance(false);
    }

    public static void reset() {
        getInstance(true);
    }

    public static DIY getInstance(boolean refresh) {
        return null;
    }


    public static void load(String classPath) {

    }

    public static void load(Class<?> root) {
    }

    public static Object getBean(String name) {
        return null;
    }
}

File: src/main/kotlin/no/utgdev/diy/DIY.kt
package no.utgdev.diy

// TODO Oppgave 2
class DIY {
    private var instantiated: Boolean = false
    private var beans: MutableSet<BeanDefinition> = mutableSetOf()
    private var objects: Map<String, Any> = mapOf()

    companion object {
        private var instance: DIY? = null

        @JvmStatic
        fun getInstance(): DIY = getInstance(false)

        @JvmStatic
        fun reset() {
            getInstance(true)
        }

        @JvmStatic
        fun getInstance(refresh: Boolean): DIY {
            if (instance == null || refresh) {
                instance = DIY()
            }
            return requireNotNull(instance)
        }

        @JvmStatic
        fun load(classPath: String) {
            getInstance().beans.addAll(DIYStatic.scan(classPath))
        }

        @JvmStatic
        fun load(root: Class<*>) {
            getInstance().beans.addAll(DIYStatic.scan(root))
        }

        @JvmStatic
        fun getBean(name: String): Any? {
            val diy = getInstance()
            if (!diy.instantiated) {
                diy.objects = DIYStatic.instansiate(diy.beans)
                DIYStatic.wireup(diy.objects)
                diy.instantiated = true
            }
            return DIYStatic.getBean(diy.objects, name)
        }
    }
}